### ISSUE TITLE
Web Socket Connection - Get disconnected intermittent

### ISSUE THREAD

**Issue by claylaut** [aspnet/SignalR#2411]
I'm facing this issue that a connection is getting closed intermittently between 10sec~40sec. (error shown on client log which is attached below)

I'm not sure what we've changed lately but before it was working fine. Any ideas how to solve this issue?

We have set:
* Sticky sessions on the NGINX load balancer
* Idle timeout is set to 3600 

### NGINX ConfigMap
* proxy-read-timeout: "3600"
* proxy-send-timeout: "3600"

### Ingress NGINX Service  (Load Balancer)
* service.beta.kubernetes.io/aws-load-balancer-connection-idle-timeout: "3600"

### NGINX Ingress
* ingress.kubernetes.io/affinity: "cookie"
* ingress.kubernetes.io/session-cookie-name: "route"
* ingress.kubernetes.io/session-cookie-hash: "sha1"

## Currently we are using:
* Signalr Javascript client : `1.0.0`
* Server-Side NuGet Packages: `1.0.0`
* Signalr Server is running on Kubernetes cluster of: `4 instances`
* Kubernetes: using `ingress with ngnix`
* Operating System on the Server: `Linux`
* Operating System on the Client: `Windows`
* Host: `AWS`
* Browser on the client: `Chrome`

## Sample Client Application for quick testing:
This sample client application connects with our signalr server
https://github.com/claylaut/signalr-websocket-connection-issue

## Server logs sample:
These server logs are for one specific user getting disconnected and reconnected multiple times. 
If you need stack trace of any particular log, please do not hesitate to ask.
[server-logs-sample](https://github.com/aspnet/SignalR/files/2058511/server-logs.xlsx)

## Client sample error:
Note: Client and server logs are from different sessions, so ID's will not correlate 
![client-trace-error](https://user-images.githubusercontent.com/1999085/40786878-7851424a-64ec-11e8-9b0f-6707fd072e39.png)

**Comment by anurse** [comment]
Looks like the connection is being terminated by the proxy. Have you configured NGINX to proxy WebSockets properly? It requires specific configuration: https://www.nginx.com/blog/websocket-nginx/

**Comment by samcov** [comment]
Hummm, didn't know that about nginx, just commenting to make a note of this when we want to deploy that way.

**Comment by claylaut** [comment]
@anurse, 

This is the headers that we received from the hub, we got these from `hub`:
`((HttpConnectionContext)Context.Features[typeof(IHttpContextFeature)]).HttpContext.Request.Headers`
```
Headers received: 
Cache-Control -> no-cache,
Connection -> upgrade,
Pragma -> no-cache,
Upgrade -> websocket,
Accept-Encoding -> gzip, deflate, br,
Accept-Language -> en-US,en;q=0.9,
Cookie -> __zlcmid=mohOuPX0onJyAa; route=a7a2cbdba8e7b7b3b378565d8a825d438758a64f,
Host -> underworld.loki.com,
User-Agent -> Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/66.0.3359.181 Safari/537.36,
Origin -> file://,
X-Request-ID -> ff571cb8bafc18d6e83fb65cefc6fb0c,
X-Real-IP -> 212.56.154.60,
X-Original-Proto -> http,
X-Forwarded-Host -> underworld.loki.com,
X-Forwarded-Port -> 80,
X-Original-URI -> /common/real-time/magic?id=crHVwzsEfyu4GjHA3GQCVw,
X-Scheme -> http,
Sec-WebSocket-Version -> 13,
Sec-WebSocket-Key -> wfs3sn0wshcVpzeO15nbTA==,
Sec-WebSocket-Extensions -> permessage-deflate; client_max_window_bits,
X-Original-For -> 100.96.4.18:32884
```

We did few other changes but unfortunatly we had no luck to fix this.

Any ideas?

**Comment by anurse** [comment]
Can you try connecting directly to the server, without using nginx as a proxy and see if the connections disconnect then? If they don't, then the issue is likely in the nginx configuration. The fact that it is disconnecting reliably after a certain period indicates that nginx is probably killing the connection due to some timeout. By default, the ASP.NET Core WebSockets middleware sends a keep-alive frame every 2 minutes, so as long as the timeouts for all the proxies are larger than that (your original issue indicated `3600` which as long as the units are seconds should be fine), the connection should remain active.

Also, since you're using Ingress, have you configured it appropriately for WebSockets? https://github.com/nginxinc/kubernetes-ingress/tree/master/examples/websocket

**Comment by claylaut** [comment]
@anurse sorry for the late reply but currently I`m abroad and I have limited internet connection.
Finally, we solved this issue it was nginx reloading and dropping connections.

in a nutshell we had to add:
--enable-dynamic-configuration
--watch-namespace=$(POD_NAMESPACE)
--force-namespace-isolation=true

For more information about this fix can be found here: https://github.com/kubernetes/ingress-nginx/issues/2647

@anurse thanks again for your help.
